### ISSUE TITLE
Differences in mount vs findmount output causing idempotency issues

### ISSUE THREAD

**Issue by axelrtgs** [chef/chef#10453]
## Description
after some digging looks like the switch to findmnt may be causing the issue as we didn’t have a problem before and chef mounts the drive so its not something manually mounted. need to investigate more but not sure if anyone might have an idea on the extra section in the path shown in the findmnt output?
findmnt -rn output
`/backups //MTLNAS/Servers/backup[/backup] cifs rw,relatime`
mount output
`//MTLNAS/Servers/backup on /backups type cifs (rw,relatime)`

double checking the regex matching in the PR it looks like it doesnt account for the random [/backup] that gets outputted in the findmnt output vs mount output. fstab also doesnt have that extra notation and matches the mount output

one of my colleagues looked into it and found this recent PR and figured something might be the cause: https://github.com/chef/chef/pull/10171/files

## Chef Version
16.5

## Platform Version
Linux (CentOS/Ubuntu)

## Replication Case
Use a CIFs mount that has a different output in mount vs findmnt. mounts will keep trying to get remounted but fail because the device is already mounted.

## Client Output

```

```

**Comment by lamont-granquist** [comment]
problem seems to be the argument in square brackets here:

```
 /backups //MTLNAS/Servers/backup[/backup] cifs rw,relatime
```

offhand it looks like that is the path on the cifs server which isn't very useful, and should just be stripped out?

**Comment by lamont-granquist** [comment]
It looks like there's a pattern match in the provider which matches stuff that has square brackets but fails on this, and its not clear that it would be appropriate (since it seems like we need to match against `//MTLNAS/Servers/backup` and not `/backup` in that case, and of course there's no tests -- although there is a rubular link for the kind of moutpoint that the existing code is attempting to mount against.

i have no idea.

**Comment by tas50** [comment]
@antima-gupta Any chance you can take a look at this one along with your other mount work?

**Comment by antima-gupta** [comment]
@tas50 Sure I'll look into this, currently working on another priority stuff.

**Comment by antima-gupta** [comment]
@axelrtgs To make `mount` resource idempotent, I used `findmnt -rn` command. `findmnt -rn` list the currently mounted file systems along with target mount point, and the source device.
Since it's idempotent, so If mounts will keep trying to get remounted, it will fail because the device is already mounted.

Here is the [Regex](https://github.com/MsysTechnologiesllc/chef/blob/4e6234d500475ec772ae2f7d81bfb91067b4a12c/lib/chef/provider/mount/linux.rb#L53) matching in the PR. It checks for the random [/backup] as well. Permalink for your case : https://rubular.com/r/r0RRTY45jE74SO, which matching with the PR Regex. 

As per replication case description, it seems like you want it should not be idempotent.

If we make it as `not idempotent` so mountpoints in fstab file are mounted again in every chef-run, if you run chef-client four times you are going to have the resource mounted four times.

Please correct me If my understanding is incorrect and could you please share piece of code that is failing and logs, would be helpful to reproduce the issue?

**Comment by axelrtgs** [comment]
So the issue is that the current regexes are not matching with `/backups //MTLNAS/Servers/backup[/backup] cifs rw,relatime` causing it to not be idempotent.

The goal of this issue is to fix the regex so it can be idempotent again. Right now even though it is already mounted on every chef run it has the unwanted behaviour of trying to mount it again resulting in an error that the device is busy.

the expected behaviour would be that after the initial mount on subsequent chef runs it recognizes the mount already exists and does nothing.

the only regex that comes close to resembling the pattern above is the one here https://github.com/chef/chef/pull/10171/files#diff-9e68c237a560753a94569d68a9f375e17612cdba759ae777d96a46960e548e2dR52 since it has a similar look but it also does not match. the regex should be something like

`when /\A#{Regexp.escape(real_mount_point)}\s+#{device_mount_regex}\s+\[[/\w]+\]/`

rubular example https://rubular.com/r/jx7iX97t8Cer3c

**Comment by antima-gupta** [comment]
@axelrtgs could you please provide more information such as recipe and error logs to reproduce the issue?

**Comment by axelrtgs** [comment]
```
mount '/backups' do
  device      '//MTLNAS/Servers/backup'
  device_type :device
  fstype      'cifs'
  fsck_device '-'
  options     [
    'iocharset=utf8',
    'file_mode=0775',
    'dir_mode=0775',
  ]
  dump        0
  pass        0

  action [:mount, :enable]
end
```
error
```
STDERR: mount error(16): Device or resource busy
Refer to the mount.cifs(8) manual page (e.g. man mount.cifs)
```

**Comment by axelrtgs** [comment]
if you simplify the resource to its basics the error still occurs.

```
mount '/backups' do
  device      '//MTLNAS/Servers/backup'
  fstype      'cifs'

  action [:mount, :enable]
end
```

**Comment by tas50** [comment]
I believe we have this fixed by #10614. I'll let you know when we resolve our windows build issues and get this out to the current channel. It would be great to know if this resolves all your issues or if we need to spec out a few more scenarios.
